modules/src/applications/transfer/context.rs

```diff
@@ -0,0 +1,348 @@
+    ) -> Result<<Self as Ics20Reader>::AccountId, Ics20Error> {
+        // https://github.com/cosmos/cosmos-sdk/blob/master/docs/architecture/adr-028-public-key-addresses.md
+        let contents = format!("{}/{}", port_id, channel_id);
+
+        let mut hasher = Sha256::new();
+        hasher.update(VERSION.as_bytes());
+        hasher.update(b"0");
+        hasher.update(contents.as_bytes());
+        let hash: Vec<u8> = hasher.finalize().to_vec().drain(20..).collect();
+
+        String::from_utf8(hex::encode_upper(hash))
+            .expect("hex encoded bytes are not valid UTF8")
+            .parse::<Signer>()
+            .map_err(Ics20Error::signer)?
+            .try_into()
+            .map_err(|_| Ics20Error::parse_account_failure())
+    }
+
+    /// Returns true iff send is enabled.
+    fn is_send_enabled(&self) -> bool;
+
+    /// Returns true iff receive is enabled.
+    fn is_receive_enabled(&self) -> bool;
+
+    /// Returns true iff the store contains a `DenomTrace` entry for the specified `HashedDenom`.
+    fn has_denom_trace(&self, hashed_denom: &HashedDenom) -> bool {
+        self.get_denom_trace(hashed_denom).is_some()
+    }
+
+    /// Get the denom trace associated with the specified hash in the store.
+    fn get_denom_trace(&self, denom_hash: &HashedDenom) -> Option<DenomTrace>;
+}
+
+pub trait BankKeeper {
+    type AccountId: ToString;
+
+    /// This function should enable sending ibc fungible tokens from one account to another
+    fn send_coins(
+        &mut self,
+        from: &Self::AccountId,
+        to: &Self::AccountId,
+        amt: &IbcCoin,
+    ) -> Result<(), Ics20Error>;
+
+    /// This function to enable minting ibc tokens in a module
+    fn mint_coins(&mut self, module: &Self::AccountId, amt: &IbcCoin) -> Result<(), Ics20Error>;
+
+    /// This function should enable burning of minted tokens
+    fn burn_coins(&mut self, module: &Self::AccountId, amt: &IbcCoin) -> Result<(), Ics20Error>;
+
+    /// This function should enable transfer of tokens from the ibc module to an account
+    fn send_coins_from_module_to_account(
+        &mut self,
+        module: &Self::AccountId,
+        to: &Self::AccountId,
+        amt: &IbcCoin,
+    ) -> Result<(), Ics20Error>;
+
+    /// This function should enable transfer of tokens from an account to the ibc module
+    fn send_coins_from_account_to_module(
+        &mut self,
+        from: &Self::AccountId,
+        module: &Self::AccountId,
+        amt: &IbcCoin,
+    ) -> Result<(), Ics20Error>;
+}
+
+pub trait BankReader {
+    type AccountId: ToString + TryFrom<Signer>;
+
+    /// Returns true if the specified account is not allowed to receive funds and false otherwise.
+    fn is_blocked_account(&self, account: &Self::AccountId) -> bool;
+
+    /// get_transfer_account returns the ICS20 - transfers AccountId.
+    fn get_transfer_account(&self) -> Self::AccountId;
+}
+
+pub trait AccountReader {
+    type AccountId: ToString;
```

Comment:
What are the benefits of using `AccountId` to identify accounts as opposed to using `Address` directly? One consequence of the current interface is that an account can have multiple addresses. Is this how we intend to think about what an "account" is?

Note: the SDK's [`AccountI`](https://github.com/cosmos/cosmos-sdk/blob/450cd7fc8708ccb0fa21f05e251d9804a2063b79/x/auth/types/account.go#L307) models an account as having only one associated address.

Reply:
Just realized that this code is no longer used after recent refactoring, so removed it as well. https://github.com/informalsystems/ibc-rs/pull/1989/commits/2d3a224d304780110290761878f4ce5246db4311